File: frontend/src/api/governance.ts
import { apiRequest } from './client'
import type { RevisionListResponse } from './articles'

export type ReviewerApplication = {
  id: number
  user_id: number
  username: string
  statement: string
  status: 'pending' | 'approved' | 'rejected'
  reviewed_by_id: number | null
  reviewed_by_name: string | null
  review_note: string
  created_at: string
  updated_at: string
  reviewed_at: string | null
}

export type ReviewerApplicationList = {
  items: ReviewerApplication[]
  total: number
}

export const reviewerApplicationKeys = {
  all: ['reviewer-applications'] as const,
  mine: ['reviewer-applications', 'mine'] as const,
  pending: ['reviewer-applications', 'pending'] as const,
}

export type AdminArticle = {
  id: number
  name: string
  description: string
  is_published: boolean
  is_taken_down: boolean
}

export type AdminArticleList = {
  items: AdminArticle[]
  total: number
}

export const governanceKeys = {
  articles: ['governance', 'articles'] as const,
  revisions: (symptomId: number) =>
    ['governance', 'articles', symptomId, 'revisions'] as const,
}

export function getMyReviewerApplication(signal?: AbortSignal) {
  return apiRequest<ReviewerApplication | null>(
    '/reviewer-applications/me',
    { signal },
  )
}

export function applyForReviewer(statement: string) {
  return apiRequest<ReviewerApplication>('/reviewer-applications', {
    method: 'POST',
    body: { statement },
  })
}

export function listReviewerApplications(signal?: AbortSignal) {
  return apiRequest<ReviewerApplicationList>(
    '/admin/reviewer-applications?status=pending',
    { signal },
  )
}

export function decideReviewerApplication(
  applicationId: number,
  action: 'approve' | 'reject',
  note: string,
) {
  return apiRequest<ReviewerApplication>(
    `/admin/reviewer-applications/${applicationId}/${action}`,
    {
      method: 'POST',
      body: { note },
    },
  )
}

export function listAdminArticles(signal?: AbortSignal) {
  return apiRequest<AdminArticleList>('/admin/articles', { signal })
}

export function listAdminRevisions(
  symptomId: number,
  signal?: AbortSignal,
) {
  return apiRequest<RevisionListResponse>(
    `/admin/articles/${symptomId}/revisions`,
    { signal },
  )
}

File: backend/app/schemas/governance.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.article import ArticleRevisionItem


class ReviewerApplicationCreate(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    statement: str = Field(min_length=10, max_length=1000)


class GovernanceDecision(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    note: str = Field(default="", max_length=1000)


class GovernanceReason(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    reason: str = Field(min_length=3, max_length=1000)


class ReviewerApplicationItem(BaseModel):
    id: int
    user_id: int
    username: str
    statement: str
    status: Literal["pending", "approved", "rejected"]
    reviewed_by_id: int | None
    reviewed_by_name: str | None
    review_note: str
    created_at: datetime
    updated_at: datetime
    reviewed_at: datetime | None


class ReviewerApplicationList(BaseModel):
    items: list[ReviewerApplicationItem]
    total: int


class AdminArticleItem(BaseModel):
    id: int
    name: str
    description: str
    is_published: bool
    is_taken_down: bool


class AdminArticleList(BaseModel):
    items: list[AdminArticleItem]
    total: int


class AdminArticleState(BaseModel):
    id: int
    is_published: bool


class AdminRevisionList(BaseModel):
    items: list[ArticleRevisionItem]
    total: int


class AuditLogItem(BaseModel):
    id: int
    actor_id: int | None
    actor_name: str | None
    action: str
    target_type: str
    target_id: str
    details: dict
    created_at: datetime


class AuditLogList(BaseModel):
    items: list[AuditLogItem]
    total: int
